Adding DynamoDB High Availability Features

kmg-stripe opened the pull request:
### Context

This is a forked version of the branch for [Adding DynamoDB high availability features](https://github.com/Netflix/mantis/pull/675)

The motivation for this PR is to ensure we can build a snapshot using the original branch, which was failing since the branch is in a fork of Mantis.  Context is here: https://github.com/Netflix/mantis/pull/675#issuecomment-2173895730 

### Checklist

- [X] `./gradlew build` compiles code correctly
- [X] Added new tests where applicable
- [X] `./gradlew test` passes all tests
- [X] Extended README or added javadocs where applicable

kmg-stripe commented:
@Andyz26 - I pulled Mike's branch into my forked repo, pushed the branch and created this PR to test building a snapshot.  It looks like I need an approval to start the build.

Andyz26 commented:
> @Andyz26 - I pulled @mcowgill-stripe's branch into my forked repo, pushed the branch and created this PR to test building a snapshot. It looks like I need an approval to start the build.

what did you see when you click the details/approval link on the "publish snapshot to ..." task above? I thought the collab/write permission groups give you permission to approve there.
BTW I've approved the build.

kmg-stripe commented:
> > @Andyz26 - I pulled @mcowgill-stripe's branch into my forked repo, pushed the branch and created this PR to test building a snapshot. It looks like I need an approval to start the build.
> 
> what did you see when you click the details/approval link on the "publish snapshot to ..." task above? I thought the collab/write permission groups give you permission to approve there. BTW I've approved the build.

@Andyz26 - Ah, I do not see anything there right now, so assume I missed it.  I'll self-approve next time, assuming I see the approval link.

It looks like the build finished successfully.  Based on the comments in #675 , it looks like you wanted to do some testing on the snapshots before we merge it?  I also see some comments in #675  .  Are you ok with addressing those in follow-on PR(s)?

Andyz26 commented:
> > > @Andyz26 - I pulled @mcowgill-stripe's branch into my forked repo, pushed the branch and created this PR to test building a snapshot. It looks like I need an approval to start the build.
> > 
> > 
> > what did you see when you click the details/approval link on the "publish snapshot to ..." task above? I thought the collab/write permission groups give you permission to approve there. BTW I've approved the build.
> 
> @Andyz26 - Ah, I do not see anything there right now, so assume I missed it. I'll self-approve next time, assuming I see the approval link.
> 
> It looks like the build finished successfully. Based on the comments in #675 , it looks like you wanted to do some testing on the snapshots before we merge it? I also see some comments in #675 . Are you ok with addressing those in follow-on PR(s)?

yeah I need to try this build to see if anything is broken. And yeah for major refactor comments/discusion thread they could defer to follow up PRs.

Andyz26 commented on mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:
```diff
@@ -346,20 +288,6 @@ public MasterConfiguration getConfig() {
-        }
-    }
-
-    public Observable<MasterDescription> getMasterObservable() {
```
looks like we still need to expose this from the new interface as we need it for some internal connections.

Andyz26 replied:
"Observable<MasterDescription>" from the leader client.

kmg-stripe replied:
Just to be clear, do you mean expose this from `ILeadershipManager` or another interface?

Andyz26 replied:
we still need "public Observable<MasterDescription> getMasterObservable()" method from teh masterMain class.

kmg-stripe replied:
ah, understood!  i'll push up a commit later today.

kmg-stripe replied:
@Andyz26 I pushed up a commit a few hours ago.

Also, I do not see anyway for me to approve snapshot creation.  I just see: "This branch is waiting to be deployed" with "Integrate Pull Request" with a link to the build.

Andyz26 replied:
just approved. I guess there is another security setting somewhere I need to find.

kmg-stripe replied:
hey @Andyz26 - Just checking in to see if you found any more issues :)

Andyz26 replied:
I got the control plane tested with some config changes and it seems to be working now. I will tried the agent/runtime parts today.

Andyz26 commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/CoreConfiguration.java:
```diff
@@ -77,4 +77,8 @@ public interface CoreConfiguration {
+
+    @Config("mantis.leader.monitor.factory")
+    @Default("io.mantisrx.server.core.master.LocalLeaderFactory")
+    String getLeaderMonitorFactory();
```
nit: getLeaderMonitorFactoryName

Andyz26 commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/ILeadershipManager.java:
```diff
@@ -0,0 +1,54 @@
+ */
+package io.mantisrx.server.core;
+/*
+ * Copyright 2019 Netflix, Inc.
```
dupe copyright?

Andyz26 commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java:
```diff
@@ -36,36 +37,45 @@
-    public void startUp() throws Exception {
+    public void start()  {
+        try {
+            if(curator.getCurator().getState() != CuratorFrameworkState.STARTED) {
```
just checking: is this thread safe?

kmg-stripe replied:
ah, yeah, looks like it is not.  it does seem that there is the potential for racing "starts" to result in a thrown exception here.

i think a small refactor can fix this.  i'll do that today.

kmg-stripe replied:
oh wait, i take that back, concurrent calls to "start()" should be safe.  subsequent calls should be no-ops.

Andyz26 commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java:
```diff
@@ -0,0 +1,151 @@
+
+public class DynamoDBLeaderElector extends BaseService {
+
+    private static final Logger logger = LoggerFactory.getLogger(DynamoDBLeaderElector.class);
```
nit: use "@Slf4j"

Andyz26 commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:
```diff
@@ -90,10 +83,11 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
-            setupCuratorListener();
-            curator.start();
-            masterMonitor.startAsync().awaitRunning();
+            if(curator.getState() != CuratorFrameworkState.STARTED) {
```
nit: add logging

kmg-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:
```diff
@@ -90,10 +83,11 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
-            isConnectedGauge.set(0L);
-            setupCuratorListener();
-            curator.start();
-            masterMonitor.startAsync().awaitRunning();
```
@Andyz26 - re: thread safety: wdyt of wrapping this in `synchronized (this) { ... }` to synchronize concurrent calls to `start()`?

kmg-stripe commented on mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBStore.java:
```diff
@@ -115,7 +115,7 @@ public Map<String, String> getAll(String tableName, String partitionKey) throws
         Map<String, AttributeValue> expressionAttributeValues = new HashMap<>();
         expressionAttributeValues.put(PK_V, AttributeValue.builder().s(tableName).build());
-        expressionAttributeValues.put(SK_V, AttributeValue.builder().s(partitionKey).build());
+        expressionAttributeValues.put(SK_V, AttributeValue.builder().s(String.format("%s#", partitionKey)).build());
```
This is unrelated to the refactor, but a bug found while testing

kmg-stripe commented on mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:
```diff
@@ -83,10 +83,13 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
-                isConnectedGauge.set(0L);
-                setupCuratorListener();
-                curator.start();
+            synchronized (this) {
```
@Andyz26 I think this should handle the potential race when start() can be called more than once (your comment on thread safety).  lmk if you think this is not correct, or can think of a better way...

kmg-stripe commented:
@Andyz26 - i addressed your comments and fixed a small, unrelated bug in the dynamodb store.  lmk if there is anything else.  otherwise, i'll merge :)
